File: LaboratoryWork/Semester_3/Laboratory_1/C++/src/ds/Stack.cpp
#include "Stack.hpp"
using namespace std;

void Stack::Init() {
    head = nullptr;
}

void Stack::Push(string obj) {
    StackNode* newNode = new StackNode{obj, head};
    head = newNode;
}

string Stack::Pop() {
    if (!head) {
        cerr << "Pop: stack is empty!" << endl;
        return "";
    }

    StackNode* removedNode = head;
    string removedValue = head->data;

    head = head->next;
    delete removedNode;
    return removedValue;
}

void Stack::Print() {
    StackNode* currentNode = head;
    while (currentNode) {
        cout << currentNode->data << " ";
        currentNode = currentNode->next;
    }
    cout << endl;
}

int Stack::GetSize() {
    StackNode* currentNode = head;
    int currentSize = 0;
    while (currentNode) {
        currentNode = currentNode->next;
        currentSize++;
    }
    return currentSize;
}

string Stack::GetElm(int pos) {
    StackNode* currentNode = head;
    int currentPos = 0;
    while (currentNode) {
        if (currentPos == pos) {
            return currentNode->data;
        }
        currentNode = currentNode->next;
        currentPos++;
    }
    throw invalid_argument("GetElm: invalid position!");
}

void Stack::Destroy() {
    StackNode* currentNode = head;
    while (currentNode) {
        StackNode* deletedNode = currentNode;
        currentNode = currentNode->next;
        delete deletedNode;
    }
    head = nullptr;
}

Stack::~Stack() {
    Destroy();
}

File: LaboratoryWork/Semester_3/Laboratory_2/GoLang/ds/stack.go
package ds

import (
	"errors"
	"fmt"
)

type StackNode[T any] struct {
	Key  T
	Next *StackNode[T]
}

type Stack[T any] struct {
	head *StackNode[T]
}

func NewStack[T any]() *Stack[T] {
	return &Stack[T]{head: nil}
}

func (s *Stack[T]) Push(value T) {
	newNode := &StackNode[T]{Key: value, Next: s.head}
	s.head = newNode
}

func (s *Stack[T]) Pop() error {
	if s.head == nil {
		return errors.New("[ERROR] Stack is empty")
	}
	s.head = s.head.Next
	return nil
}

func (s *Stack[T]) Top() (T, error) {
	if s.head == nil {
		var zero T
		return zero, errors.New("[ERROR] Stack is empty")
	}
	return s.head.Key, nil
}

func (s *Stack[T]) Empty() bool {
	return s.head == nil
}

func (s *Stack[T]) Print() {
	for node := s.head; node != nil; node = node.Next {
		fmt.Printf("%v ", node.Key)
	}
	fmt.Println()
}